lib/dry/result_matcher/either_matcher.rb

```diff
@@ -0,0 +1,28 @@
+require "dry/result_matcher/case"
+
+module Dry
+  module ResultMatcher
+    EitherMatcher = Dry::ResultMatcher::Matcher.new(
+      success: Case.new(
+        match: -> result, *pattern {
+          result = result.to_either if result.respond_to?(:to_either)
+          result.right?
+        },
+        resolve: -> result {
+          result = result.to_either if result.respond_to?(:to_either)
+          result.value
+        },
+      ),
+      failure: Case.new(
+        match: -> result, *pattern {
+          result = result.to_either if result.respond_to?(:to_either)
+          result.left?
+        },
+        resolve: -> result {
+          result = result.to_either if result.respond_to?(:to_either)
+          result.value
+        },
+      )
```

Comment:
They're so we can work with objects that aren't already eithers, but can _convert_ to eithers, for example, Dry::Monads::Try success/failure objects.

> On 29 Jun 2016, at 7:31 AM, Piotr Solnica <notifications@github.com> wrote:
> 
> In lib/dry/result_matcher/either_matcher.rb:
> 
> > +        },
> > +        resolve: -> result {
> > +          result = result.to_either if result.respond_to?(:to_either)
> > +          result.value
> > +        },
> > +      ),
> > +      failure: Case.new(
> > +        match: -> result, *pattern {
> > +          result = result.to_either if result.respond_to?(:to_either)
> > +          result.left?
> > +        },
> > +        resolve: -> result {
> > +          result = result.to_either if result.respond_to?(:to_either)
> > +          result.value
> > +        },
> > +      )
> these respond_to calls shouldn't be needed, why do we have them?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>